compiler/test/contextFreeGrammar.test.ts

```diff
@@ -0,0 +1,90 @@
+import { it } from "vitest";
+import { describe } from "vitest";
+
+import { ContextFreeGrammar } from "../dist";
+import { beforeEach } from "vitest";
+import { expect } from "vitest";
+import { SimplifiedGrammarRepresentation } from "../dist/utils";
+import {
+  findFirstSets,
+  findFollowSets,
+} from "../dist/context-free-grammar/algorithms/top-down-parsing";
+
+describe("Testing Compiler", () => {
+  let grammar: ContextFreeGrammar;
+
+  beforeEach(() => {
+    const program = `
+    S -> 0 1 S | 1 0 S | A.
+    A -> 0 1 A | 1 0 A | #.
+      `;
+    grammar = new ContextFreeGrammar(program);
+  });
+  it("Parsing", () => {
+    grammar.parse();
+    expect(grammar.errors).toHaveLength(0);
+    expect(grammar.parseTree).toBeDefined();
+    expect(grammar.result).toBeDefined();
+  });
+
+  it("Semantic Analysis", () => {
+    grammar.parse().semanticAnalysis();
+    expect(grammar.errors).toHaveLength(0);
+    expect(grammar.warnings).toHaveLength(0);
+    expect(grammar["grammar"]).toBeInstanceOf(SimplifiedGrammarRepresentation);
+  });
+
+  it("Find First Sets", () => {
+    grammar.parse().semanticAnalysis();
+    const firstSets = grammar.findFirstSets();
+    expect(firstSets).toBeDefined();
+    expect(firstSets).toEqual(findFirstSets(grammar["grammar"]));
```

Comment:
This is not the right way of defining tests. It's like checking if result of function is equal to itself(which will always be true!). We should rather test for correctness of the functions by comparing the output to an output we know is correct for the given grammar.